File: pump/eopatch/src/main/java/info/nightscout/androidaps/plugins/pump/eopatch/ble/task/FetchAlarmTask.java
package info.nightscout.androidaps.plugins.pump.eopatch.ble.task;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

import info.nightscout.androidaps.plugins.pump.eopatch.alarm.IAlarmRegistry;
import info.nightscout.androidaps.plugins.pump.eopatch.core.api.GetErrorCodes;
import info.nightscout.androidaps.plugins.pump.eopatch.core.response.AeCodeResponse;
import info.nightscout.rx.bus.RxBus;
import info.nightscout.rx.logging.LTag;
import io.reactivex.rxjava3.core.Single;

@Singleton
public class FetchAlarmTask extends TaskBase {
    @Inject RxBus rxBus;
    @Inject IAlarmRegistry alarmRegistry;

    private final GetErrorCodes ALARM_ALERT_ERROR_CODE_GET;

    @Inject
    public FetchAlarmTask() {
        super(TaskFunc.FETCH_ALARM);
        ALARM_ALERT_ERROR_CODE_GET = new GetErrorCodes();
    }

    public Single<AeCodeResponse> getPatchAlarm() {
        return isReady()
                .concatMapSingle(v -> ALARM_ALERT_ERROR_CODE_GET.get())
                .doOnNext(this::checkResponse)
                .firstOrError()
                .doOnSuccess(aeCodeResponse -> alarmRegistry.add(aeCodeResponse.getAlarmCodes()))
                .doOnError(e -> aapsLogger.error(LTag.PUMPCOMM, (e.getMessage() != null) ? e.getMessage() : "FetchAlarmTask error"));
    }

    public synchronized void enqueue() {
        boolean ready = (disposable == null || disposable.isDisposed());

        if (ready) {
            disposable = getPatchAlarm()
                .timeout(TASK_ENQUEUE_TIME_OUT, TimeUnit.SECONDS)
                .subscribe();
        }
    }
}

File: pump/eopatch/src/main/kotlin/app/aaps/pump/eopatch/ble/task/FetchAlarmTask.kt
package app.aaps.pump.eopatch.ble.task

import app.aaps.core.interfaces.logging.LTag
import app.aaps.pump.eopatch.alarm.IAlarmRegistry
import app.aaps.pump.eopatch.core.api.GetErrorCodes
import app.aaps.pump.eopatch.core.response.AeCodeResponse
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.functions.Consumer
import io.reactivex.rxjava3.functions.Function
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Suppress("PrivatePropertyName")
@Singleton
class FetchAlarmTask @Inject constructor(
    private val alarmRegistry: IAlarmRegistry
) : TaskBase(TaskFunc.FETCH_ALARM) {

    private val ALARM_ALERT_ERROR_CODE_GET: GetErrorCodes = GetErrorCodes()

    fun getPatchAlarm(): Single<AeCodeResponse> {
        return isReady()
            .concatMapSingle<AeCodeResponse>(Function { ALARM_ALERT_ERROR_CODE_GET.get() })
            .doOnNext(Consumer { response: AeCodeResponse -> this.checkResponse(response) })
            .firstOrError()
            .doOnSuccess(Consumer { aeCodeResponse: AeCodeResponse -> alarmRegistry.add(aeCodeResponse.alarmCodes) })
            .doOnError(Consumer { e: Throwable -> aapsLogger.error(LTag.PUMPCOMM, e.message ?: "FetchAlarmTask error") })
    }

    @Synchronized override fun enqueue() {
        val ready = (disposable == null || disposable?.isDisposed == true)

        if (ready) {
            disposable = getPatchAlarm()
                .timeout(TASK_ENQUEUE_TIME_OUT, TimeUnit.SECONDS)
                .subscribe()
        }
    }
}
